File: clj/src/slipstream/ui/views/section.clj
(ns slipstream.ui.views.section
  (:require [net.cgrand.enlive-html :as html]
            [slipstream.ui.util.core :as u]
            [slipstream.ui.util.clojure :as uc]
            [slipstream.ui.util.enlive :as ue]
            [slipstream.ui.util.icons :as icons]
            [slipstream.ui.views.subsection :as subsection]))

(def template-filename (u/template-path-for "section.html"))

(def section-selected-cls "ss-section-selected")

(def section-group-sel [:#ss-section-group])
(def section-sel [[:.ss-section (ue/first-of-class "ss-section")]])
(def section-flat-sel [[:.ss-section (ue/first-of-class "ss-section-flat")]])
(def section-anchor-sel (concat section-sel [:.ss-section-activator]))
(def chevron-sel (concat section-anchor-sel [:span [:span (html/nth-child 2)]]))
(def section-panel-sel [[:.panel-collapse :.collapse]])
(def section-header-sel [:.ss-section-header])
(def section-icon-sel [:.ss-section-icon])
(def section-title-sel [:.ss-section-title])
(def section-content-sel [:.ss-section-content])



(html/defsnippet section-flat-snip template-filename section-flat-sel
  []
  identity)

(html/defsnippet section-group-snip template-filename section-group-sel
  [total-section-count section-group-index sections]
  ue/this (ue/append-to-id "-" section-group-index)
  ue/this (ue/content-for section-sel [{:keys [icon title selected? content type] :as section} sections
                   :let [section-uid (->> (or title "untitled") uc/keywordize name (str "ss-section-" (.indexOf sections section) "-"))
                         unique-section? (= total-section-count 1)
                         collapsible?    (not unique-section?)
                         flat-section?   (-> section :type (= :flat-section))]]
    ue/this             (ue/when-substitute flat-section? (section-flat-snip))
    section-sel         (ue/enable-class (uc/first-not-nil selected? unique-section?) section-selected-cls)
    section-sel         (ue/when-add-class type (str "ss-section-" (name type)))
    section-icon-sel    (when icon (icons/set (icon :tooltip-placement "bottom")))
    section-title-sel   (html/html-content (str title))
    section-header-sel  (ue/remove-if-not title)
    section-content-sel (ue/if-enlive-node content
                          (if (string? content)
                            (html/html-content content)
                            (html/content content))
                          (subsection/build content))
    section-anchor-sel  (ue/if-set-href collapsible?
                                       (str "#" section-uid)
                                       nil)
    section-anchor-sel  (ue/set-data :parent ".panel-group")
    chevron-sel         (ue/remove-if-not collapsible?)
    section-panel-sel   (ue/set-id section-uid)))

(defn build
  [sections]
  (let [relevant-sections   (remove nil? sections)
        total-section-count (count relevant-sections)]
    (html/content (->> relevant-sections
                       u/ensure-one-selected
                       (partition-by (comp :section-group meta))
                       (map-indexed (partial section-group-snip total-section-count))
                       concat))))

File: clj/src/slipstream/ui/views/js/section.js
jQuery( function() { ( function( $$, $, undefined ) {

    var hash = document.location.hash,
        sectionIdPrefix = "ss-section-";

    if (hash) {
        // Try to open section from url hash
        var hashSegments = hash.trimPrefix("#")
                                .split($$.util.url.hash.segmentSeparator)
                                .filter($$.util.string.notEmpty),
            sectionTitle = hashSegments[0],
            subSectionTitle = hashSegments[1];
            $section = $('.panel #' + sectionIdPrefix + sectionTitle);
        if ($section.foundOne()) {
            $section.addClass("in");
            if (subSectionTitle) {
                // Try to open a subsection
                if (! $$.subsections.showByTitle($section, subSectionTitle)) {
                    // There is no subsection with the give title, so we clean it.
                    document.location.hash = sectionTitle;
                }
            }
        } else {
            // There is no section with the give id (or more than one), so we clean it.
            $$.util.url.reloadPageWithoutHashInURL();
        }
    } else {
        // Open panel of section by default
        $(".panel.ss-section-selected .panel-collapse.collapse").addClass("in");
    }

    // Ensure correct chevrons at page load depending on open/close state

    $(".panel .panel-collapse.collapse.in")
        .parent()
        .find(".panel-title .glyphicon-chevron-down")
        .removeClass("glyphicon-chevron-down")
        .addClass("glyphicon-chevron-up");

    $(".panel .panel-collapse.collapse:not(.in)")
        .parent()
        .find(".panel-title .glyphicon-chevron-up")
        .removeClass("glyphicon-chevron-up")
        .addClass("glyphicon-chevron-down");


    // Ensure correct chevrons and hash when opening/closing the sections

    $(".panel-collapse.collapse").on("show.bs.collapse", function (e) {
        var chevron_sel = ".panel-title a[href='#" + e.delegateTarget.id + "'] .glyphicon";
        $(chevron_sel).removeClass('glyphicon-chevron-down').addClass('glyphicon-chevron-up');
        // Update hash in URL
        window.location.hash = e.delegateTarget.id.trimPrefix(sectionIdPrefix);
    });

    $(".panel-collapse.collapse").on("hide.bs.collapse", function (e) {
        var chevron_sel = ".panel-title a[href='#" + e.delegateTarget.id + "'] .glyphicon";
        $(chevron_sel).removeClass('glyphicon-chevron-up').addClass('glyphicon-chevron-down');
        if (window.location.hash.trimPrefix("#") == e.delegateTarget.id.trimPrefix(sectionIdPrefix)) {
            // Remove the hash from URL if it contains still the hash of the closing section
            window.location.hash = "";
        }
    });

    if ($$.util.meta.isPageType("edit new")) {
        // For pages in edit mode, bring the focus to the first textfield of the first open section
        $(".panel .panel-collapse.collapse.in").focusFirstInput();

        $(".panel-collapse.collapse").on("shown.bs.collapse", function (e) {
            // And ensure the focus each time when a section is open.
            $(this).focusFirstInput();
        });
    }



    // Update .ss-panel-{info,danger,warning,success} to Bootstrap classes

    $("div.panel.ss-section-info")
        .removeClass("panel-default")
        .addClass("panel-info");

    $("div.panel.ss-section-danger")
        .removeClass("panel-default")
        .addClass("panel-danger");

    $("div.panel.ss-section-warning")
        .removeClass("panel-default")
        .addClass("panel-warning");

    $("div.panel.ss-section-success")
        .removeClass("panel-default")
        .addClass("panel-success");

    function toggleCollapsible(bsPanel) {
        var $bsPanelToOpen = $(bsPanel);

        if (! $bsPanelToOpen.length){
            return "nothing to select";
        }

        if ($bsPanelToOpen.find(".collapse.in").length){
            return "already selected";
        }

        $bsPanelToOpen
            .closest(".panel-group")
            .find(".collapse.in")
            .collapse("hide");

        $bsPanelToOpen
            .find(".collapse")
            .collapse("show");

        return "sucessfully selected";
    }

    $$.Section = {
        select: function (index) {
            return toggleCollapsible(
                $(".panel-group .panel:nth-child(" + index + ")")
                    .closest(".panel")
            );
        },
        selectByTitle: function (title) {
            return toggleCollapsible(
                $(".ss-section-title")
                    .filter(function(){return $(this).text().trim() == title.trim();})
                    .closest(".panel")
            );
        }
    };

}( window.SlipStream = window.SlipStream || {}, jQuery ));});
